File: Scala/src/main/scala/AlgorithmicHeights/SAT2.scala
package AlgorithmicHeights

object SAT2 {
  import scala.annotation.tailrec
  import scala.collection.mutable.ListBuffer

  private type Component = List[Node]
  final case class DFSResult(components: List[Component],
                             topologicalOrdering: List[Node],
                             previsitIds: List[Int],
                             postvisitIds: List[Int])
  final case class Node(originalId: Int, nrNodes: Int) {
    override def toString: String = s"$originalId"

    val nodeId: Int = if (originalId < 0) -originalId + nrNodes else originalId

    def unary_- : Node = Node(-originalId, nrNodes)
  }
  final case class Edge(a: Node, b: Node)

  class DirectedGraph(val nrNodes: Int, val edgeList: List[Edge], nodeOrder: Option[List[Node]] = None) {
    import DirectedGraph.buildAdjacencyList

    val orderedNodes: List[Node] = nodeOrder match {
      case None => (1 to nrNodes).map(k => Node(originalId(k), nrNodes / 2)).toList
      case Some(nodes) => nodes
    }

    private val adjacencyList: Map[Node, List[Node]] = buildAdjacencyList(edgeList)

    def neighbours(node: Node): List[Node] = adjacencyList.getOrElse(node, Nil)

    private def originalId(nodeId: Int): Int = {
      val n: Int = nrNodes / 2
      if (nodeId <= n) nodeId else -(nodeId - n)
    }
  }

  object DirectedGraph {
    private def buildAdjacencyList(edgeList: List[Edge]): Map[Node, List[Node]] =
      edgeList.groupMap{ case Edge(a, _) => a }{ case Edge(_, b) => b }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readParameters(line: String): (Int, Int) = convertToIntList(line) match {
    case List(n, e) => (n, e)
    case _ => throw new Exception("Malformed input.")
  }

  private[AlgorithmicHeights] def readEdges(reader: Iterator[String], nrNodes: Int, nrEdges: Int): List[Edge] =
    reader
      .take(nrEdges)
      .map(convertToIntList)
      .collect{ case List(a, b) => (a, b) }
      .flatMap{
        case (a, b) =>
          val node1 = Node(a, nrNodes)
          val node2 = Node(b, nrNodes)
          List(Edge(-node1, node2), Edge(-node2, node1))
      }
      .toList

  private def readGraphs(reader: Iterator[String], n: Int): List[DirectedGraph] = {
    (0 until n).map{ _ =>
      val _: String = reader.next()
      val (nrNodes, nrEdges): (Int, Int) = readParameters(reader.next())
      val edgeList: List[Edge] = readEdges(reader, nrNodes, nrEdges)
      new DirectedGraph(2*nrNodes, edgeList)
    }.toList
  }

  private def depthFirstSearch(graph: DirectedGraph): DFSResult = {
    val visitStarted: Array[Int] = Array.fill(graph.nrNodes)(0)
    val visitEnded: Array[Int] = Array.fill(graph.nrNodes)(0)
    val topologicalSorting: ListBuffer[Node] = ListBuffer()
    val previsitId: Iterator[Int] = Iterator.from(1)
    val postvisitId: Iterator[Int] = Iterator.from(1)

    def isNodeVisited(node: Node): Boolean = visitStarted(node.nodeId - 1) > 0

    def findUnvisitedNeighbour(node: Node): Option[Node] = graph.neighbours(node).find(!isNodeVisited(_))

    def explore(startingNode: Node): Component = {
      @tailrec
      def traverseComponent(previsitStack: List[Node], component: Component): Component = previsitStack match {
        case Nil => component
        case node :: restOfStack =>
          findUnvisitedNeighbour(node) match {
            case Some(neighbour) =>
              visitStarted(neighbour.nodeId - 1) = previsitId.next()
              traverseComponent(neighbour :: previsitStack, neighbour :: component)
            case None =>
              visitEnded(node.nodeId - 1) = postvisitId.next()
              topologicalSorting += node
              traverseComponent(restOfStack, component)
          }
      }

      visitStarted(startingNode.nodeId - 1) = previsitId.next()
      traverseComponent(List(startingNode), List(startingNode))
    }

    @tailrec
    def findComponents(nodes: List[Node], components: List[Component]): List[Component] = nodes match {
      case Nil => components
      case node :: remainingNodes =>
        if (isNodeVisited(node)) findComponents(remainingNodes, components)
        else {
          val currentComponent: Component = explore(node)
          findComponents(remainingNodes, currentComponent :: components)
        }
    }

    val components: List[Component] = findComponents(graph.orderedNodes, Nil)
    DFSResult(components, topologicalSorting.reverseIterator.toList, visitStarted.toList, visitEnded.toList)
  }

  private def createGraphWithEdgesReversed(graph: DirectedGraph, postvisitIds: List[Int]): DirectedGraph = {
    val reversedEdges: List[Edge] = graph.edgeList.map{ case Edge(from, to) => Edge(to, from) }
    val nodeOrder: List[Node] =
      graph.orderedNodes
        .zip(postvisitIds)
        .sortBy{ case (_, postvisitId) => postvisitId }(Ordering[Int].reverse)
        .map{ case (node, _) => node }
    new DirectedGraph(graph.nrNodes, reversedEdges, Some(nodeOrder))
  }

  private def calcStronglyConnectedComponents(graph: DirectedGraph): List[Component] = {
    val DFSResult(_, _, _, postvisitNumbers) = depthFirstSearch(graph)
    val reversedGraph: DirectedGraph = createGraphWithEdgesReversed(graph, postvisitNumbers)
    val DFSResult(components, _, _, _) = depthFirstSearch(reversedGraph)
    components
  }

  private def calcComponentAssignment(component: Component, literalAssignment: Set[Node]): Option[Set[Node]] = {
    @tailrec
    def loop(acc: Set[Node], nodes: List[Node]): Option[Set[Node]] = nodes match {
      case Nil => Some(acc)
      case node :: rest =>
        if (acc.contains(-node)) None
        else if (!literalAssignment.contains(node) && !literalAssignment.contains(-node)) loop(acc + node, rest)
        else loop(acc, rest)
    }

    loop(Set.empty[Node], component)
  }

  def solve2SATProblem(graph: DirectedGraph): Set[Node] = {
    @tailrec
    def loop(acc: Set[Node], cs: List[Component]): Set[Node] = cs match {
      case Nil => acc
      case component :: css =>
        calcComponentAssignment(component, acc) match {
          case None => Set()
          case Some(componentAssignment) => loop(acc ++ componentAssignment, css)
        }
    }

    val components: List[Component] = calcStronglyConnectedComponents(graph)
    loop(Set.empty[Node], components)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrExamples: Int = reader.next().toInt
    val graphs: List[DirectedGraph] = readGraphs(reader, nrExamples)
    val result: List[Set[Node]] = graphs.map(solve2SATProblem)
    result.foreach{
      line =>
        if (line.isEmpty) println("0")
        else {
          val sortedResult: List[Node] = line.toList.sortBy{ node => math.abs(node.originalId) }
          println(s"1 ${sortedResult.mkString(" ")}")
        }
    }
  }
}

File: OCaml/src/algorithmic_heights/sat2.ml
module Node = struct
    type t = { original_id : int; nr_nodes : int; node_id : int }

    let compare {original_id = id1; _ } { original_id = id2; _ } =
        compare id1 id2

    let create original_id nr_nodes =
        let node_id = if original_id < 0 then -original_id + nr_nodes else original_id in
        { original_id; nr_nodes; node_id }

    let negate { original_id; nr_nodes; _ } =
        create (-original_id) nr_nodes

    let string_of_node { original_id; _ } =
        string_of_int original_id
end

module NodeMap = Map.Make(Node)

type edge = { a : Node.t; b : Node.t }
type component = Node.t list
type dfsresult = { components : component list; previsit_ids : int list; postvisit_ids : int list }

module Graph : sig
    type t
    val dfs : t -> dfsresult
    val create : int -> edge list -> Node.t list option -> t
    val node_order : t -> Node.t list
end = struct
    type t = { nr_nodes : int; adjacency_list : Node.t list NodeMap.t; ordered_nodes : Node.t list }

    let node_order { ordered_nodes; _ } = ordered_nodes

    let get_original_id nr_nodes node_id =
        let n = nr_nodes / 2 in
        if node_id <= n then node_id else -(node_id - n)

    let get_neighbors adjacency_list node =
        match NodeMap.find_opt node adjacency_list with
            | None -> []
            | Some neighbors -> neighbors

    let create nr_nodes edges node_order =
        let add_edge adjacency_list { a; b } =
            let update_function = function
                | None -> Some [b]
                | Some ns -> Some (b :: ns) in
            NodeMap.update a update_function adjacency_list in
        let adjacency_list = List.fold_left add_edge NodeMap.empty edges in
        let ordered_nodes = match node_order with
            | None -> List.init nr_nodes (fun k -> Node.create (get_original_id nr_nodes (k + 1)) (nr_nodes / 2))
            | Some nodes -> nodes in
        { nr_nodes; adjacency_list; ordered_nodes }

    let dfs { nr_nodes; adjacency_list; ordered_nodes } =
        let visit_started = Array.make nr_nodes 0 in
        let visit_ended = Array.make nr_nodes 0 in
        let previsit_id = ref 1 in
        let postvisit_id = ref 1 in
        let is_node_visited (node : Node.t) = visit_started.(node.node_id - 1) > 0 in
        let find_unvisited_neighbor node =
            node |> get_neighbors adjacency_list |> List.find_opt (fun item -> not (is_node_visited item)) in

        let explore (start_node : Node.t) =
            let rec traverse_component component = function
                | [] -> component
                | (node :: rest) as previsit_stack ->
                    match find_unvisited_neighbor node with
                        | Some neighbor ->
                            visit_started.(neighbor.node_id - 1) <- !previsit_id;
                            previsit_id := !previsit_id + 1;
                            traverse_component (neighbor :: component) (neighbor :: previsit_stack)
                        | None ->
                            visit_ended.(node.node_id - 1) <- !postvisit_id;
                            postvisit_id := !postvisit_id + 1;
                            traverse_component component rest in
            visit_started.(start_node.node_id - 1) <- !previsit_id;
            previsit_id := !previsit_id + 1;
            traverse_component [start_node] [start_node] in

        let rec find_components components = function
            | [] -> components
            | node :: remaining_nodes ->
                if is_node_visited node then find_components components remaining_nodes
                else
                    let current_component = explore node in
                    find_components (current_component :: components) remaining_nodes in

        let components = find_components [] ordered_nodes in
        { components; previsit_ids = Array.to_list visit_started; postvisit_ids = Array.to_list visit_ended }
end


let read_edges (nr_nodes : int) (nr_edges : int) : edge list =
    let parse_line line =
        let create_edges a b =
            let node1 = Node.create a nr_nodes in
            let node2 = Node.create b nr_nodes in
            [{ a = Node.negate node1; b = node2 }; { a = Node.negate node2; b = node1 }] in
        Scanf.sscanf line "%d %d" (fun a b -> List.to_seq (create_edges a b)) in
    let range = Seq.init nr_edges (fun k -> k) in
    List.of_seq @@ Seq.concat_map (fun _ -> parse_line (read_line ())) range


let read_graphs (n : int) : (int * edge list) list =
    let read_graph_data () =
        ignore (read_line ());
        let (nr_nodes, nr_edges) = Scanf.sscanf (read_line ()) "%d %d" (fun n e -> (n, e)) in
        let edges = read_edges nr_nodes nr_edges in
        (2*nr_nodes, edges) in
    List.(map (fun _ -> read_graph_data ()) @@ init n (fun _ -> 0))


let get_node_visit_order_in_reversed_graph (postvisit_ids : int list) (node_order : Node.t list) : Node.t list =
    let compare_pairs (_, id1) (_, id2) = compare id2 id1 in
    List.(postvisit_ids |> combine node_order |> sort compare_pairs |> map (fun (node, _) -> node ))


let calc_strongly_connected_components (nr_nodes : int) (edges : edge list) : component list =
    let graph = Graph.create nr_nodes edges None in
    let { postvisit_ids; _ } = Graph.dfs graph in
    let reversed_edges = List.map (fun {a; b} -> {a = b; b = a}) edges in
    let node_order = get_node_visit_order_in_reversed_graph postvisit_ids (Graph.node_order graph) in
    let reversed_graph = Graph.create nr_nodes reversed_edges (Some node_order) in
    let { components; _ } = Graph.dfs reversed_graph in
    components


let calc_component_assignment (component : component) (literal_assignment : Node.t list) : Node.t list option =
    let rec loop acc = function
        | [] -> Some acc
        | node :: rest ->
            if List.mem (Node.negate node) acc then None
            else if not (List.mem node literal_assignment) && not (List.mem (Node.negate node) literal_assignment) then
                loop (node :: acc) rest
            else loop acc rest in
    loop [] component


let solve_2sat (nr_nodes : int) (edges : edge list) : Node.t list =
    let rec loop acc = function
        | [] -> acc
        | component :: css ->
            match calc_component_assignment component acc with
                | None -> []
                | Some component_assignment -> loop (acc @ component_assignment) css in
    let components = calc_strongly_connected_components nr_nodes edges in
    loop [] components


let display_result (nodes : Node.t list) : unit =
    if nodes = [] then print_endline "0"
    else
        let compare_nodes (node1 : Node.t) (node2 : Node.t) =
            compare (abs node1.original_id) (abs node2.original_id) in
        let node_strings = List.(nodes |> sort compare_nodes |> map Node.string_of_node) in
        print_endline @@ String.concat " " ("1" :: node_strings)


let () =
    let nr_examples = read_int () in
    let graph_data = read_graphs nr_examples in
    let results = List.map (fun (nr_nodes, edges) -> solve_2sat nr_nodes edges) graph_data in
    List.iter display_result results
